template/src/rtm-events/EventUtils.ts

```diff
@@ -0,0 +1,182 @@
+        return i;
+      }
+    }
+
+    return -1;
+  };
+
+  return {
+    getEvents(source: EventSourceEnum) {
+      return _events[source] || (_events = {});
+    },
+    addListener(evt: string, listener: any, source: EventSourceEnum) {
+      console.log('debugging addListener', evt, listener, source);
+      if (!_isValidListener(listener)) {
+        throw new Error('Listener must be a function');
+      }
+      const listeners = _getListenersAsObject(evt, source);
+      const listenerIsWrapped = typeof listener === 'object';
+      for (let key in listeners) {
+        if (
+          listeners.hasOwnProperty(key) &&
+          _indexOfListener(listeners[key], listener) === -1
+        ) {
+          listeners[key].push(
+            listenerIsWrapped
+              ? listener
+              : {
+                  listener: listener,
+                  once: false,
+                },
+          );
+        }
+      }
+      return this;
+    },
+    removeEvent(evt: string, source: EventSourceEnum) {
+      let type = typeof evt;
+      let events = this.getEvents(source);
+      // Remove different things depending on the state of evt
+      if (type === 'string') {
+        // Remove all listeners for the specified event
+        // events[source].delete(evt);
+      }
+      return this;
+    },
+    emit(evt: string, args: any) {
+      if (!evt) {
+        console.log('emitting: event name is empty');
+        return;
+      }
+      let listenersMap = _getListenersAsObject(
+        evt,
+        args.payload.source || EventSourceEnum.core,
+      );
+      console.log('CUSTOM_EVENT_API emit listenersMap: ', listenersMap);
+      let listeners;
+      let listener;
+      for (let key in listenersMap) {
+        if (listenersMap.hasOwnProperty(key)) {
+          listeners = listenersMap[key].slice(0);
+          for (let i = 0; i < listeners.length; i++) {
+            // If the listener returns true then it shall be removed from the event
+            // The function is executed either with a basic call or an apply if there is an args array
+            listener = listeners[i];
+            if (listener.once === true) {
+              EventUtils.removeListener(
+                evt,
+                args.payload.source,
+                listener.listener,
+              );
+            }
+            const newargs = [].slice.call(arguments, 1);
+            console.log('CUSTOM_EVENT_API newargs: ', newargs);
+            listener.listener.apply(this, newargs || []);
+          }
+        }
+      }
+    },
+    removeListener(
+      evt: string,
+      source: EventSourceEnum,
+      listenerToRemove: IListener,
+    ) {
+      let listeners = _getListenersAsObject(evt, source);
+      for (let key in listeners) {
+        if (listeners.hasOwnProperty(key)) {
+          let index = _indexOfListener(listeners[key], listenerToRemove);
+          if (index !== -1) {
+            listeners[key].splice(index, 1);
```

Comment:
avoid modifying the original array, just override with a new one, splice modifies the original one